plugin_tests/segmentation_label_test.py

```diff
@@ -153,3 +155,157 @@ def test_trace_boundary(self):
+
+        # refenece neighbors
+        rx = [2, 2, 2]
+        ry = [2, 3, 2]
+
+        # test E
+        m_neighbor = np.array([[0, 0, 0, 0, 0],
+                               [0, 0, 0, 0, 0],
+                               [0, 0, 1, 1, 0],
+                               [0, 0, 0, 0, 0],
+                               [0, 0, 0, 0, 0]], dtype=np.bool)
+        x = 2
+        y = 2
+
+        by, bx = trace_boundary(m_neighbor, Connectivity=8,
+                                XStart=y, YStart=x)
+
+        np.testing.assert_allclose(rx, bx)
+        np.testing.assert_allclose(ry, by)
+
+        # refenece neighbors
+        rx = [2, 3, 2]
+        ry = [2, 3, 2]
+
+        # test SE
+        m_neighbor = np.array([[0, 0, 0, 0, 0],
+                               [0, 0, 0, 0, 0],
+                               [0, 0, 1, 0, 0],
+                               [0, 0, 0, 1, 0],
+                               [0, 0, 0, 0, 0]], dtype=np.bool)
+        x = 2
+        y = 2
+
+        by, bx = trace_boundary(m_neighbor, Connectivity=8,
+                                XStart=y, YStart=x)
+
+        np.testing.assert_allclose(rx, bx)
+        np.testing.assert_allclose(ry, by)
+
+        # refenece neighbors
+        rx = [2, 3, 2]
+        ry = [2, 2, 2]
+
+        # test S
+        m_neighbor = np.array([[0, 0, 0, 0, 0],
+                               [0, 0, 0, 0, 0],
+                               [0, 0, 1, 0, 0],
+                               [0, 0, 1, 0, 0],
+                               [0, 0, 0, 0, 0]], dtype=np.bool)
+        x = 2
+        y = 2
+
+        by, bx = trace_boundary(m_neighbor, Connectivity=8,
+                                XStart=y, YStart=x)
+
+        np.testing.assert_allclose(rx, bx)
+        np.testing.assert_allclose(ry, by)
+
+        # refenece neighbors
+        rx = [2, 3, 2]
+        ry = [2, 1, 2]
+
+        # test SW
+        m_neighbor = np.array([[0, 0, 0, 0, 0],
+                               [0, 0, 0, 0, 0],
+                               [0, 0, 1, 0, 0],
+                               [0, 1, 0, 0, 0],
+                               [0, 0, 0, 0, 0]], dtype=np.bool)
+        x = 2
+        y = 2
+
+        by, bx = trace_boundary(m_neighbor, Connectivity=8,
+                                XStart=y, YStart=x)
+
+        np.testing.assert_allclose(rx, bx)
+        np.testing.assert_allclose(ry, by)
+
+        # refenece neighbors
+        rx = [2, 2, 2]
+        ry = [2, 1, 2]
+
+        # test W
+        m_neighbor = np.array([[0, 0, 0, 0, 0],
+                               [0, 0, 0, 0, 0],
+                               [0, 1, 1, 0, 0],
+                               [0, 0, 0, 0, 0],
+                               [0, 0, 0, 0, 0]], dtype=np.bool)
+        x = 2
+        y = 2
+
+        by, bx = trace_boundary(m_neighbor, Connectivity=8,
+                                XStart=y, YStart=x)
+
+        np.testing.assert_allclose(rx, bx)
+        np.testing.assert_allclose(ry, by)
```

Comment:
It does not work because the Moore neighbor tracing needs to be back at the start point. For example, if there are two points (0,0), (0,1) and the start point (0,0), it will be back when beginning so that (0,-1) will cause an error. Also, isbf algorithm also has a condition to be considered. I think we should use a padding array before using it.